### PR TITLE
WFS3 MVT, GeoJSON tiles and tilingSchemes Implementation

### PR DESCRIPTION
**Pull request by fernandor777**
This PR implements the Mapbox vector tile output format for the “collections/{collectionId}/items” WFS 3 endpoint, tilingSchemes list output and getTiles endpoint.
Also adds a "resolution" integer vendor parameter (256 by default) for to reach smallers features that aren't printed with default tile resolution.

Get items example:
```
http://localhost:8080/geoserver/wfs3/collections/topp__solar/items?f=application%2Fx-protobuf%3Btype%3Dmapbox-vector&resolution=10000
```
Get tiling schemes list example:
```
http://localhost:8080/geoserver/wfs3/tilingSchemes
```
Get tile example:
```
http://localhost:8080/geoserver/wfs3/collections/st__solar/tiles/GlobalCRS84Geometric/12/2055/2318
```

### CODE REVIEW

**Comment by bradh** [comment]
This is failing format checks on Travis CI. Can you please take a look and fix the problems (or just do a maven build locally)?

**Comment by fernandor777** [comment]
Tested and rebased, pushed again. Thanks.

**Comment by bradh** [comment]
Format checks are still failing

```
[INFO] Processed 31 files (1 non-complying).
[ERROR] Found 1 non-complying files, failing build
[ERROR] To fix formatting errors, run "mvn fmt:format"
[ERROR] Non complying file: /home/travis/build/geoserver/geoserver/src/community/wps-remote/src/main/java/org/geoserver/wps/remote/plugin/output/XMPPRawDataOutput.java
```
and
```
[INFO] Processed 9 files (1 non-complying).
[ERROR] Found 1 non-complying files, failing build
[ERROR] To fix formatting errors, run "mvn fmt:format"
[ERROR] Non complying file: /home/travis/build/geoserver/geoserver/src/community/security/keycloak/src/main/java/org/geoserver/security/keycloak/GeoServerKeycloakFilter.java
```

File: src/community/wfs3/src/main/java/org/geoserver/wfs3/response/GetFeatureMapboxOutputFormat.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                final String layerName = feature.getName().getLocalPart();
+                final Map<String, Object> properties = getProperties(feature);
+                // add to encoder (if have at least a coordinate to encode)
+                if (finalGeom.getCoordinates().length > 0)
+                    encoder.addFeature(layerName, properties, finalGeom);
+            }
+            // write encoded stream
+            output.write(encoder.encode());
+        } catch (Exception e) {
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (simboss) [reply src/community/wfs3/src/main/java/org/geoserver/wfs3/response/GetFeatureMapboxOutputFormat.java]: I second Andrea. PrintStackTrace bypasses the whole logging machinery.

Simply put, never use it, IMHO.

[the remark this answers is not in the document]
Comment (fernandor777) [reply src/community/wfs3/src/main/java/org/geoserver/wfs3/response/GetFeatureMapboxOutputFormat.java]: My mistake, I forgot to replace the IDE autogenerated catch fill

File: src/community/wfs3/src/main/java/org/geoserver/wfs3/response/GetFeatureMapboxOutputFormat.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+import org.opengis.feature.Attribute;
+import org.opengis.feature.ComplexAttribute;
+import org.opengis.feature.GeometryAttribute;
+import org.opengis.feature.Property;
+import org.opengis.feature.simple.SimpleFeature;
+import org.opengis.referencing.FactoryException;
+import org.opengis.referencing.crs.CoordinateReferenceSystem;
+
+/** Mapbox protobuf WFS3 output format */
+public class GetFeatureMapboxOutputFormat extends WFSGetFeatureOutputFormat {
```
[the remark this answers is not in the document]
Comment (fernandor777) [reply src/community/wfs3/src/main/java/org/geoserver/wfs3/response/GetFeatureMapboxOutputFormat.java]: GeoJson is now supported by the current GeoJson output format, about TopoJSON I'll need to check its implementation and how it works, as far I remember every output format in Vector Tiles modules has its own impl class and behavior.

**Comment by fernandor777** [comment]
Also sent a GeoWebCache PR for promoving a used method in GridSet class from protected to public.

https://github.com/GeoWebCache/geowebcache/pull/689

File: src/community/wfs3/src/main/java/org/geoserver/wfs3/response/GetFeatureMapboxOutputFormat.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (aaime) [reply src/community/wfs3/src/main/java/org/geoserver/wfs3/response/GetFeatureMapboxOutputFormat.java]: Are you sure GeoJSON tiles generated by the vector tiles module are exactly the same as GeoJSON WFS output? I'd be surprised if they were.

File: src/community/wfs3/src/main/java/org/geoserver/wfs3/response/GetFeatureMapboxOutputFormat.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+                collection
+                        .getSchema()
+                        .getGeometryDescriptor()
+                        .getType()
+                        .getCoordinateReferenceSystem();
+        ReferencedEnvelope area = ReferencedEnvelope.create(collection.getBounds(), refSys);
+        // Build the Pipeline (sort of coordinates transformer)
+        Pipeline pipeline = getPipeline(area, paintArea, refSys, resolution / 32);
+        // setup the vector tile encoder
+        VectorTileEncoder encoder = new VectorTileEncoderNoClip(resolution, resolution / 32, false);
```
[the remark this answers is not in the document]
Comment (aaime) [reply src/community/wfs3/src/main/java/org/geoserver/wfs3/response/GetFeatureMapboxOutputFormat.java]: Good

**Comment by aaime** [comment]
Taking this one over, will do the residual fixes and merge directly

**Pull request closed without merging**